### ISSUE TITLE
GraphQL Transform - Searching a @searchable @model with @auth does not scope results to current user

### ISSUE THREAD

**Issue by gabehollombe-aws** [aws-amplify/amplify-cli#460]
**Describe the bug**
Using Amplify with AppSync and a GraphQL Transform, I want to have a `@model` protected by `@auth` so that users can only work with items they own. This all works fine, but if I add `@searchable` to the type and then use the `searchFoo` query, the results returned are not scoped to the current user performing the query. 

**To Reproduce**

1. Set up Auth in an Amplify app using a Cognito User Pool

2. Set up a model in the schema.graphql file like this:

```
type Note @model @auth(rules: [{allow: owner}]) @searchable {
    id: ID!
    content: String!
}
```

3. Run `amplify push`

4. Create two users confirmed in the Cognito Pool, UserA and UserB

5. Auth to AppSync with UserA and issue a createNote mutation

6. Auth to AppSync with UserB and issue a searchNote query searching for content that matches what you created with UserA.  The content will be returned, but I don't expect it should be, because of the @auth directive on the Note type.

For more explicit instructions, follow Ed Lima's blog post here: https://medium.com/open-graphql/create-a-multiuser-graphql-crud-l-app-in-10-minutes-with-the-new-aws-amplify-cli-and-in-a-few-73aef3d49545 

**Expected behavior**
I expect that when I search using the `searchFoo` query for a type `Foo` with @model @auth and @searchable directives, I'll only get results back that are owned by the currently authenticated user.

**Additional context**
I'm seeing this using Amplify 0.1.34

**Comment by mikeparisstuff** [comment]
Hello @gabehollombe-aws. Unfortunately, auth support for search has yet to be implemented but is near the top of our todo list. A team member will be looking at this soon to solve for this use case. Thanks for your note. We will reference this issue when the feature is in PR.

**Comment by YikSanChan** [comment]
FYI, there seems to be a duplicate issue and possible temporary solution: https://github.com/aws-amplify/amplify-js/issues/2113

**Comment by YikSanChan** [comment]
@mikeparisstuff Is there a rough timeline for this feature? Thanks!

> auth support for search has yet to be implemented but is near the top of our todo list.

**Comment by YikSanChan** [comment]
@gabehollombe-aws I provide a solution with code sample in aws-amplify/amplify-js#2113. Feel free to check!

**Comment by kaustavghosh06** [comment]
@gabehollombe-aws We support custom resolvers now and this could be implemented as a part of the custom resolvers as @YikSanChan mentioned. He has provided the resolver code for this in https://github.com/aws-amplify/amplify-js/issues/2113. Please feel free to re-open if you think this is still an issue.

**Comment by gabehollombe-aws** [comment]
@kaustavghosh06   This is still an issue. Yes customer resolvers can solve this, but this report is about how the default behavior of @auth and @searchable should act together.  It's being discussed in https://github.com/aws-amplify/amplify-cli/issues/1043  so I think this issue should remain open until Proposal 2 mentioned in the above issue is addressed.

**Comment by cmcnierney** [comment]
In case anyone else runs into this issue and is looking for a workaround until the default behavior is changed (if it's decided to be changed), here's an [example](https://github.com/cmcnierney/current-issues/tree/master/aws-amplify/inject-auth-into-search) injecting the user context into the request instead of filtering the response - hopefully it saves someone else a little time.

**Comment by sacrampton** [comment]
After quite a bit of time I've managed to get the static & dynamic auth working in ElasticSearch. The posts above were all very helpful in getting there, but I thought I'd share my findings in case it saves someone else a bit of time.

Firstly, I followed the approach of @YikSanChan above, but decided that doing it in the response was not the right approach for me as it retrieves all records from the database and you filter them in the response rather than the request. Also, the total / count reflects all records retrieved from the database rather than the auth filter.

So then I followed the approach from @cmcnierney above which modifies the request so you only retrieve the records of interest. This was a lot more difficult because I couldn't find any documentation on $util.transform.toElasticsearchQueryDSL($ctx.args.filter). You need to understand how this works so you can ensure that you are making a valid ElasticSearch Query. What worked for me was to insert invalid JSON at the start of the request (ie. DEBUG { "version": "2017-02-28") so that it fails on execution, but it displays the JSON it is trying to send to ElasticSearch and you can test in Kibana Console whether that is what you are trying to achieve and progressively debug it.

As I mentioned, $util.transform.toElasticsearchQueryDSL($ctx.args.filter) is a bit of a black box and it behaves differently depending on the number of filters in your appsync query. So if we have 2 appsync queries:
- QUERY 1
`query searchAssets{
    searchAssets(filter:{
       name: {eq: "PA-101"}})}`
- QUERY 2
`query searchAssets{
    searchAssets(filter:{
       name: {eq: "PA-101"}
       description: {matchPhrase: "PUMP"}})}`

When we feed this to $util.transform.toElasticsearchQueryDSL($ctx.args.filter) it returns the following for each query:
- QUERY 1
`{"match": {"name": "PA-101"}}

- QUERY 2
`{ "bool": {
      "must": [
        {"match": {"name": "PA-101"}},
        {"match_phrase" : {"description": "PUMP"}}
      ]
    }
  }`

So if you directly inject this into the query in the example it will fail. For my purpose I had to check how many filters are in the query and if there was more than 1 strip the last couple of characters that were returned off so it would construct a valid query.

Finally, the query I used to simulate the auth was essentially

`{ "bool": {
       "minimum_should_match": 1,
       "should": [
      
        In this section user your groups from Cognito (can see how to get that from from
        amplify generated req.vtl) the do a match with your group field. Setting the
        "minimum_should_match to 1 is essential to ensure that a least one of the groups in
        Cognito matches your group field.
        {"match: {"groupfield": "${userGroup}"}

       ],
      "must": [
        {"match": {"name": "PA-101"}},
        {"match_phrase" : {"description": "PUMP"}}
      ]
    }
}`

Using Kibana Console to ensure I was constructing a valid ElasticSearch query was essential. Good luck.

**Comment by SwaySway** [comment]
Closing this issue as @auth on @searchable support was included in the 3.13 release.
